src/tools/miri/src/bin/miri.rs

```diff
@@ -109,11 +109,14 @@ impl rustc_driver::Callbacks for MiriBeRustCompilerCalls {
                 // an empty result if `tcx.sess.opts.output_types.should_codegen()` is false.
                 local_providers.exported_symbols = |tcx, cnum| {
                     assert_eq!(cnum, LOCAL_CRATE);
+                    let reachable_set = tcx.with_stable_hashing_context(|hcx| {
+                        tcx.reachable_set(()).to_sorted(&hcx, true)
+                    });
                     tcx.arena.alloc_from_iter(
                         // This is based on:
                         // https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L62-L63
                         // https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L174
-                        tcx.reachable_set(()).iter().filter_map(|&local_def_id| {
+                        reachable_set.into_iter().filter_map(|&local_def_id| {
```

Comment:
> The only place where additional sorting is happening is in Miri's override implementation of the exported_symbols query.

This was copied from elsewhere in rustc, so if the change only happens in Miri but not the source where we copied this from, that seems suspicious?

Reply:
The return type of [reachable_non_generics_provider](https://github.com/rust-lang/rust/blob/8b1dbf728add722d4db894b9b986ec24e1cdb0a1/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L42) is a `DefIdMap`, i.e. a type that does not have a visible iteration order, so no sorting needs to happen there.

rustc's version of `exported_symbols` already does some stable sorting before returning the slice, so that is fine too. That Miri didn't sort the slice before returning could be considered a bug because queries must be deterministic across compilation sessions (and the iteration order of a HashSet of DefId/DefIndex/LocalDefId isn't). It's unlikely that this can cause a problem if incremental compilation isn't involved though.